## NBA Basketball - Offensive and Defensive Game Play

In [1]:
from pymongo import MongoClient
import pprint

import pandas as pd
import matplotlib.pyplot as plt
plt.style.use('fivethirtyeight')

# Requests sends and receives HTTP requests.
import requests

import json
import time
import copy

In [2]:
from nba_api.stats.static import teams
from nba_api.stats.static import players
from nba_api.stats.endpoints import leaguegamelog
from nba_api.stats.endpoints import boxscorefourfactorsv2
from nba_api.stats.endpoints import leaguedashteamstats

In [3]:
teams_df = pd.DataFrame(teams.get_teams())
teams_df

id               full_name abbreviation       nickname  \
0   1610612737           Atlanta Hawks          ATL          Hawks   
1   1610612738          Boston Celtics          BOS        Celtics   
2   1610612739     Cleveland Cavaliers          CLE      Cavaliers   
3   1610612740    New Orleans Pelicans          NOP       Pelicans   
4   1610612741           Chicago Bulls          CHI          Bulls   
5   1610612742        Dallas Mavericks          DAL      Mavericks   
6   1610612743          Denver Nuggets          DEN        Nuggets   
7   1610612744   Golden State Warriors          GSW       Warriors   
8   1610612745         Houston Rockets          HOU        Rockets   
9   1610612746    Los Angeles Clippers          LAC       Clippers   
10  1610612747      Los Angeles Lakers          LAL         Lakers   
11  1610612748              Miami Heat          MIA           Heat   
12  1610612749         Milwaukee Bucks          MIL          Bucks   
13  1610612750  Minnesota Timberwolves          MIN   Timberwolves   
14  1610612751           Brooklyn Nets          BKN           Nets   
15  1610612752         New York Knicks          NYK         Knicks   
16  1610612753           Orlando Magic          ORL          Magic   
17  1610612754          Indiana Pacers          IND         Pacers   
18  1610612755      Philadelphia 76ers          PHI          76ers   
19  1610612756            Phoenix Suns          PHX           Suns   
20  1610612757  Portland Trail Blazers          POR  Trail Blazers   
21  1610612758        Sacramento Kings          SAC          Kings   
22  1610612759       San Antonio Spurs          SAS          Spurs   
23  1610612760   Oklahoma City Thunder          OKC        Thunder   
24  1610612761         Toronto Raptors          TOR        Raptors   
25  1610612762               Utah Jazz          UTA           Jazz   
26  1610612763       Memphis Grizzlies          MEM      Grizzlies   
27  1610612764      Washington Wizards          WAS        Wizards   
28  1610612765         Detroit Pistons          DET        Pistons   
29  1610612766       Charlotte Hornets          CHA        Hornets   

             city                 state  year_founded  
0         Atlanta               Atlanta          1949  
1          Boston         Massachusetts          1946  
2       Cleveland                  Ohio          1970  
3     New Orleans             Louisiana          2002  
4         Chicago              Illinois          1966  
5          Dallas                 Texas          1980  
6          Denver              Colorado          1976  
7    Golden State            California          1946  
8         Houston                 Texas          1967  
9     Los Angeles            California          1970  
10    Los Angeles            California          1948  
11          Miami               Florida          1988  
12      Milwaukee             Wisconsin          1968  
13      Minnesota             Minnesota          1989  
14       Brooklyn              New York          1976  
15       New York              New York          1946  
16        Orlando               Florida          1989  
17        Indiana               Indiana          1976  
18   Philadelphia          Pennsylvania          1949  
19        Phoenix               Arizona          1968  
20       Portland                Oregon          1970  
21     Sacramento            California          1948  
22    San Antonio                 Texas          1976  
23  Oklahoma City              Oklahoma          1967  
24        Toronto               Ontario          1995  
25           Utah                  Utah          1974  
26        Memphis             Tennessee          1995  
27     Washington  District of Columbia          1961  
28        Detroit              Michigan          1948  
29      Charlotte        North Carolina          1988

In [4]:
player_df = pd.DataFrame(players.get_players())
player_df

id            full_name first_name     last_name  is_active
0       76001       Alaa Abdelnaby       Alaa     Abdelnaby      False
1       76002      Zaid Abdul-Aziz       Zaid    Abdul-Aziz      False
2       76003  Kareem Abdul-Jabbar     Kareem  Abdul-Jabbar      False
3          51   Mahmoud Abdul-Rauf    Mahmoud    Abdul-Rauf      False
4        1505    Tariq Abdul-Wahad      Tariq   Abdul-Wahad      False
...       ...                  ...        ...           ...        ...
4496  1627790           Ante Zizic       Ante         Zizic       True
4497    78647             Jim Zoet        Jim          Zoet      False
4498    78648            Bill Zopf       Bill          Zopf      False
4499  1627826          Ivica Zubac      Ivica         Zubac       True
4500    78650           Matt Zunic       Matt         Zunic      False

[4501 rows x 5 columns]

In [5]:
#game_log_1819 = leaguegamelog.LeagueGameLog(counter = 0, direction = "DESC",
#                 league_id = "00", player_or_team_abbreviation = "T", season = "2018-19",
#                 season_type_all_star = "Regular Season", sorter = "DATE").get_data_frames()

In [6]:
#game_log_1819[0].to_csv(r'data/game_log_1819.csv', index=False)

In [7]:
game_log_1819 = pd.read_csv('../data_1/game_log_1819.csv')
game_log_1819.head()

SEASON_ID     TEAM_ID TEAM_ABBREVIATION               TEAM_NAME   GAME_ID  \
0      22018  1610612763               MEM       Memphis Grizzlies  21801225   
1      22018  1610612744               GSW   Golden State Warriors  21801225   
2      22018  1610612753               ORL           Orlando Magic  21801222   
3      22018  1610612766               CHA       Charlotte Hornets  21801222   
4      22018  1610612750               MIN  Minnesota Timberwolves  21801228   

    GAME_DATE      MATCHUP WL  MIN  FGM  ...  DREB  REB  AST  STL  BLK  TOV  \
0  2019-04-10  MEM vs. GSW  W  240   48  ...    34   49   25    9    0   11   
1  2019-04-10    GSW @ MEM  L  240   46  ...    30   39   32    3    5   14   
2  2019-04-10    ORL @ CHA  W  240   48  ...    30   39   24    5    2    6   
3  2019-04-10  CHA vs. ORL  L  240   41  ...    28   32   21    3    4    8   
4  2019-04-10    MIN @ DEN  L  240   39  ...    34   41   24    6    0   10   

   PF  PTS  PLUS_MINUS  VIDEO_AVAILABLE  
0  18  132          15                1  
1  17  117         -15                1  
2  19  122           8                1  
3  14  114          -8                1  
4  22   95          -4                1  

[5 rows x 29 columns]

In [8]:
#boxscorefourfactor = boxscorefourfactorsv2.BoxScoreFourFactorsV2(game_id='0021800997').get_data_frames()

In [9]:
#boxscorefourfactor[0].to_csv(r'data/boxscorefourfactor.csv', index=False)

In [10]:
boxscorefourfactor = pd.read_csv('../data_1/boxscorefourfactor.csv')
boxscorefourfactor.head()

GAME_ID     TEAM_ID TEAM_ABBREVIATION TEAM_CITY  PLAYER_ID   PLAYER_NAME  \
0  21800997  1610612745               HOU   Houston     201569   Eric Gordon   
1  21800997  1610612745               HOU   Houston     200782   P.J. Tucker   
2  21800997  1610612745               HOU   Houston     203991  Clint Capela   
3  21800997  1610612745               HOU   Houston     201935  James Harden   
4  21800997  1610612745               HOU   Houston     101108    Chris Paul   

  START_POSITION COMMENT    MIN  EFG_PCT  FTA_RATE  TM_TOV_PCT  OREB_PCT  \
0              F     NaN  31:48    0.432     0.220       0.099     0.286   
1              F     NaN  37:51    0.471     0.286       0.141     0.293   
2              C     NaN  35:39    0.409     0.242       0.155     0.302   
3              G     NaN  33:55    0.467     0.328       0.143     0.278   
4              G     NaN  32:50    0.452     0.242       0.148     0.282   

   OPP_EFG_PCT  OPP_FTA_RATE  OPP_TOV_PCT  OPP_OREB_PCT  
0        0.459         0.224        0.214         0.261  
1        0.441         0.339        0.206         0.345  
2        0.464         0.321        0.197         0.280  
3        0.510         0.294        0.220         0.320  
4        0.490         0.353        0.194         0.217

### General Dataset Information

In [11]:
game_log_1819.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2460 entries, 0 to 2459
Data columns (total 29 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   SEASON_ID          2460 non-null   int64  
 1   TEAM_ID            2460 non-null   int64  
 2   TEAM_ABBREVIATION  2460 non-null   object 
 3   TEAM_NAME          2460 non-null   object 
 4   GAME_ID            2460 non-null   int64  
 5   GAME_DATE          2460 non-null   object 
 6   MATCHUP            2460 non-null   object 
 7   WL                 2460 non-null   object 
 8   MIN                2460 non-null   int64  
 9   FGM                2460 non-null   int64  
 10  FGA                2460 non-null   int64  
 11  FG_PCT             2460 non-null   float64
 12  FG3M               2460 non-null   int64  
 13  FG3A               2460 non-null   int64  
 14  FG3_PCT            2460 non-null   float64
 15  FTM                2460 non-null   int64  
 16  FTA                2460 

In [12]:
game_log_1819.describe()

SEASON_ID       TEAM_ID       GAME_ID          MIN          FGM  \
count     2460.0  2.460000e+03  2.460000e+03  2460.000000  2460.000000   
mean     22018.0  1.610613e+09  2.180062e+07   241.646341    41.082114   
std          0.0  8.657201e+00  3.551425e+02     7.601350     5.219906   
min      22018.0  1.610613e+09  2.180000e+07   240.000000    25.000000   
25%      22018.0  1.610613e+09  2.180031e+07   240.000000    38.000000   
50%      22018.0  1.610613e+09  2.180062e+07   240.000000    41.000000   
75%      22018.0  1.610613e+09  2.180092e+07   240.000000    44.000000   
max      22018.0  1.610613e+09  2.180123e+07   340.000000    61.000000   

               FGA       FG_PCT         FG3M         FG3A      FG3_PCT  ...  \
count  2460.000000  2460.000000  2460.000000  2460.000000  2460.000000  ...   
mean     89.210569     0.461429    11.363821    32.008943     0.355224  ...   
std       7.332872     0.053312     3.730194     7.310012     0.086793  ...   
min      64.000000     0.278000     2.000000    12.000000     0.115000  ...   
25%      85.000000     0.426000     9.000000    27.000000     0.296000  ...   
50%      89.000000     0.460000    11.000000    32.000000     0.353000  ...   
75%      94.000000     0.495000    14.000000    37.000000     0.409250  ...   
max     123.000000     0.649000    27.000000    70.000000     0.842000  ...   

              DREB          REB          AST          STL          BLK  \
count  2460.000000  2460.000000  2460.000000  2460.000000  2460.000000   
mean     34.818293    45.165447    24.586585     7.633740     4.953252   
std       5.468226     6.692565     5.119577     2.947763     2.485716   
min      18.000000    22.000000    10.000000     0.000000     0.000000   
25%      31.000000    41.000000    21.000000     6.000000     3.000000   
50%      35.000000    45.000000    24.000000     7.000000     5.000000   
75%      38.000000    50.000000    28.000000     9.000000     6.000000   
max      55.000000    71.000000    42.000000    20.000000    19.000000   

               TOV           PF          PTS   PLUS_MINUS  VIDEO_AVAILABLE  
count  2460.000000  2460.000000  2460.000000  2460.000000      2460.000000  
mean     14.082927    20.904472   111.208537     0.000000         1.011382  
std       3.927105     4.293642    12.651393    14.659231         0.106100  
min       3.000000     9.000000    68.000000   -56.000000         1.000000  
25%      11.000000    18.000000   103.000000    -9.000000         1.000000  
50%      14.000000    21.000000   111.000000     0.000000         1.000000  
75%      17.000000    24.000000   120.000000     9.000000         1.000000  
max      29.000000    38.000000   168.000000    56.000000         2.000000  

[8 rows x 24 columns]

In [13]:
game_log_1819.duplicated().sum()

0

In [14]:
boxscorefourfactor.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25 entries, 0 to 24
Data columns (total 17 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   GAME_ID            25 non-null     int64  
 1   TEAM_ID            25 non-null     int64  
 2   TEAM_ABBREVIATION  25 non-null     object 
 3   TEAM_CITY          25 non-null     object 
 4   PLAYER_ID          25 non-null     int64  
 5   PLAYER_NAME        25 non-null     object 
 6   START_POSITION     10 non-null     object 
 7   COMMENT            6 non-null      object 
 8   MIN                19 non-null     object 
 9   EFG_PCT            19 non-null     float64
 10  FTA_RATE           19 non-null     float64
 11  TM_TOV_PCT         19 non-null     float64
 12  OREB_PCT           19 non-null     float64
 13  OPP_EFG_PCT        19 non-null     float64
 14  OPP_FTA_RATE       19 non-null     float64
 15  OPP_TOV_PCT        19 non-null     float64
 16  OPP_OREB_PCT       19 non-nu

In [15]:
boxscorefourfactor.describe()

GAME_ID       TEAM_ID     PLAYER_ID    EFG_PCT   FTA_RATE  \
count        25.0  2.500000e+01  2.500000e+01  19.000000  19.000000   
mean   21800997.0  1.610613e+09  6.267465e+05   0.454684   0.357263   
std           0.0  1.529706e+00  7.043532e+05   0.068524   0.272100   
min    21800997.0  1.610613e+09  1.717000e+03   0.333000   0.000000   
25%    21800997.0  1.610613e+09  2.015690e+05   0.418500   0.238500   
50%    21800997.0  1.610613e+09  2.030930e+05   0.450000   0.290000   
75%    21800997.0  1.610613e+09  1.627827e+06   0.475500   0.382500   
max    21800997.0  1.610613e+09  1.629151e+06   0.591000   1.333000   

       TM_TOV_PCT   OREB_PCT  OPP_EFG_PCT  OPP_FTA_RATE  OPP_TOV_PCT  \
count   19.000000  19.000000    19.000000     19.000000    19.000000   
mean     0.145789   0.208053     0.520526      0.379421     0.146737   
std      0.050902   0.125224     0.240371      0.281309     0.046124   
min      0.000000   0.000000     0.402000      0.125000     0.059000   
25%      0.126500   0.147000     0.439500      0.231000     0.118000   
50%      0.155000   0.278000     0.464000      0.294000     0.141000   
75%      0.186000   0.297500     0.495000      0.382500     0.184000   
max      0.195000   0.350000     1.500000      1.333000     0.220000   

       OPP_OREB_PCT  
count     19.000000  
mean       0.216632  
std        0.100940  
min        0.000000  
25%        0.173500  
50%        0.231000  
75%        0.279500  
max        0.345000

In [16]:
boxscorefourfactor.duplicated().sum()

0

### Dataset Preparation


### Exploratory Data Analysis

### Hypothesis Tests

**Scientific question:** Is there a play between offensive and defensive play that is more effective in achieving a game win?

**Null:** There is no difference between the impact of offensive or defensive play in achieving a game win.

**Alternative:** There is a difference between the impact of offensive or defensive play in achieving a game win.

**Test statistic:** 2,460 regular season games from the 20xx - xx NBA season
(=82 games per team * 30 NBA teams)

Dataset: using proxies, such as either offensive (ORtg) and defensive (DRtg) ratings or offense and defense four factors, for efficient offensive and defensive play respectively

### Conclusion